File: adventofcode-2016/src/main/java/nl/mvdr/adventofcode/adventofcode2016/day24/Part2.java
package nl.mvdr.adventofcode.adventofcode2016.day24;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import nl.mvdr.adventofcode.solver.Solver;

public class Part2 implements Solver {

    private static final Logger LOGGER = LoggerFactory.getLogger(Part2.class);

    @Override
    public String solve(String inputfile) {
        return new AirDuctSpelunkingSolver(true).solve(inputfile);
    }
    
    public static void main(String[] args) {
        Part2 instance = new Part2();

        String result = instance.solve("input-day24.txt");

        LOGGER.info(result);
    }
}

File: adventofcode-2025/src/main/kotlin/nl/mvdr/adventofcode/adventofcode2025/day02/Part2.kt
package nl.mvdr.adventofcode.adventofcode2025.day02

import io.github.oshai.kotlinlogging.KotlinLogging
import nl.mvdr.adventofcode.solver.FunctionSolver

private val logger = KotlinLogging.logger{}

fun solvePart2(lines: Sequence<String>): Long =
    lines.first()
        .split(",")
        .map { it -> it.split("-") }
        .map { it -> it.first().toLong() .. it.last().toLong() }
        .flatMap { it -> it.asSequence() }
        .filter(::isInvalid)
        .sum()

private fun isInvalid(id: Long): Boolean {
    val idString = id.toString()
    val length = idString.length
    return (1 ..< length)
        .filter { it -> length % it == 0 }
        .any { it -> isRepeating(idString, it) }
}

private fun isRepeating(idString: String, substringLength: Int) =
    (0..<idString.length / substringLength)
        .map { it -> idString.substring(it * substringLength, (it + 1) * substringLength) }
        .distinct()
        .count() == 1

fun main() {
    val result = FunctionSolver(::solvePart2).solve("input-day02.txt")
    logger.info { result }
}
